ee/hogai/notebook/notebook_serializer.py

```diff
@@ -0,0 +1,610 @@
+
+logger = logging.getLogger(__name__)
+
+
+class MarkdownTokenizer:
+    """Simple markdown tokenizer that handles the most common markdown elements."""
+
+    def __init__(self):
+        self.tokens = []
+        self.pos = 0
+        self.text = ""
+
+    def tokenize(self, text: str) -> list[dict]:
+        """Tokenize markdown text into a list of tokens."""
+        self.text = text
+        self.pos = 0
+        self.tokens = []
+
+        while self.pos < len(self.text):
+            if not self._try_parse_block_element():
+                # If no block element found, parse as paragraph
+                self._parse_paragraph()
+
+        return self.tokens
+
+    def _try_parse_block_element(self) -> bool:
+        """Try to parse a block-level element. Returns True if successful."""
+        # Skip empty lines
+        if self._at_line_start() and self._current_line().strip() == "":
+            self._skip_line()
+            return True
+
+        # Try different block elements
+        if self._try_parse_heading():
+            return True
+        if self._try_parse_code_block():
+            return True
+        if self._try_parse_blockquote():
+            return True
+        if self._try_parse_horizontal_rule():
+            return True
+        if self._try_parse_list():
+            return True
+
+        return False
+
+    def _try_parse_heading(self) -> bool:
+        """Parse heading (# ## ### etc)."""
+        if not self._at_line_start():
+            return False
+
+        line = self._current_line()
+        match = re.match(r"^(#{1,6})\s+(.+)$", line)
+        if match:
+            level = len(match.group(1))
+            content = match.group(2).strip()
+            self.tokens.append({"type": "heading", "level": level, "content": content})
+            self._skip_line()
+            return True
+        return False
+
+    def _try_parse_code_block(self) -> bool:
+        """Parse fenced code block (``` language)."""
+        if not self._at_line_start():
+            return False
+
+        line = self._current_line()
+        match = re.match(r"^```(\w*)\s*$", line)
+        if match:
+            language = match.group(1) or None
+            self._skip_line()
+
+            # Collect code lines until closing ```
+            code_lines = []
+            while self.pos < len(self.text):
+                line = self._current_line()
+                if line.strip() == "```":
+                    self._skip_line()
+                    break
+                code_lines.append(line)
+                self._skip_line()
+
+            self.tokens.append({"type": "code_block", "language": language, "content": "\n".join(code_lines)})
+            return True
+        return False
+
+    def _try_parse_blockquote(self) -> bool:
+        """Parse blockquote (> text)."""
+        if not self._at_line_start():
+            return False
+
+        line = self._current_line()
+        if line.startswith("> "):
```

Comment:
Can it start with `>` without a space? Would it make sense to have a regex with `\s?`?